allennlp/data/vocabulary.py

```diff
@@ -406,6 +406,113 @@ def from_params(cls, params: Params, instances: Iterable['adi.Instance'] = None)
+                    pretrained_list = None
+                token_counts = list(counter[namespace].items())
+                token_counts.sort(key=lambda x: x[1], reverse=True)
+                max_vocab = max_vocab_size[namespace]
+                if max_vocab:
+                    token_counts = token_counts[:max_vocab]
+                for token, count in token_counts:
+                    if pretrained_list is not None:
+                        if only_include_pretrained_words:
+                            if token in pretrained_list and count >= min_count.get(namespace, 1):
+                                self.add_token_to_namespace(token, namespace)
+                        elif token in pretrained_list or count >= min_count.get(namespace, 1):
+                            self.add_token_to_namespace(token, namespace)
+                    elif count >= min_count.get(namespace, 1):
+                        self.add_token_to_namespace(token, namespace)
+
+        if tokens_to_add:
+            for namespace, tokens in tokens_to_add.items():
+                for token in tokens:
+                    self.add_token_to_namespace(token, namespace)
+
+    def extend_from_instances(self,
+                              params: Params,
+                              instances: Iterable['adi.Instance'] = ()) -> None:
+        """
+        This method can be used for extending already generated vocabulary
+        from collection of instances. It is a wrapped around ``extend`` method.
+        """
+        min_count = params.pop("min_count", None)
+        max_vocab_size = params.pop_int("max_vocab_size", None)
+        non_padded_namespaces = params.pop("non_padded_namespaces", DEFAULT_NON_PADDED_NAMESPACES)
+        pretrained_files = params.pop("pretrained_files", {})
+        only_include_pretrained_words = params.pop_bool("only_include_pretrained_words", False)
+        tokens_to_add = params.pop("tokens_to_add", None)
+        params.assert_empty("Vocabulary - from dataset")
+
+        logger.info("Fitting token dictionary from dataset.")
+        namespace_token_counts: Dict[str, Dict[str, int]] = defaultdict(lambda: defaultdict(int))
+        for instance in Tqdm.tqdm(instances):
+            instance.count_vocab_items(namespace_token_counts)
+        self.extend(counter=namespace_token_counts,
+                    min_count=min_count,
+                    max_vocab_size=max_vocab_size,
+                    non_padded_namespaces=non_padded_namespaces,
+                    pretrained_files=pretrained_files,
+                    only_include_pretrained_words=only_include_pretrained_words,
+                    tokens_to_add=tokens_to_add)
+
+    def add_non_padded_namespaces(self, non_padded_namespaces: Set[str]):
+        self._token_to_index.add_non_padded_namespaces(non_padded_namespaces)
+        self._index_to_token.add_non_padded_namespaces(non_padded_namespaces)
+        self._non_padded_namespaces.update(non_padded_namespaces)
```

Comment:
Will be called internally only: `Set[str]` vs `Iterable[str]` and `set(..)` call?